File: Algorithms/T787_Cheapest_Flights_Kstop.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
        List<int[]>[] graph = new LinkedList[n];
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        for(int[] edge : flights){
            if(graph[edge[0]] == null) graph[edge[0]] = new LinkedList();
            if(graph[edge[1]] == null) graph[edge[1]] = new LinkedList();
            graph[edge[0]].add(new int[]{edge[1], edge[2]});
        }
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> (a[1] - b[1]));
        pq.add(new int[]{src, 0, K});
        while(!pq.isEmpty()){
            int[] cur = pq.poll();
            if(cur[0] == dst) return cur[1];
            if(cur[2] >= 0){
                for(int[] nei : graph[cur[0]]){
                    if(dist[nei[0]] > cur[1] + nei[1]) pq.add(new int[]{nei[0], cur[1] + nei[1], cur[2] - 1});
                }
            }
        }
        return -1;
    }
}

File: Algorithms/C++/T787_Cheapest_Flights_K_Stop.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        queue<pair<int, int>> q;
        vector<vector<pair<int, int>>> graph(n, vector<pair<int, int>>());
        vector<int> map(n, INT_MAX);
        for(auto f : flights){
            graph[f[0]].push_back(pair(f[1], f[2]));
        }
        q.push(pair(src, 0));
        map[src] = 0;
        while(!q.empty() && K >= 0){
            int num = q.size();
            for(int i = 0; i < num; ++i){
                auto cur = q.front();   
                q.pop();
                for(auto p : graph[cur.first]){                
                    int dist = p.second + cur.second;
                    if(p.first == dst){
                        map[dst] = min(map[dst], dist);
                    }else if(dist < map[p.first]){
                        q.push(pair(p.first, dist));
                        map[p.first] = dist;
                    }
                }              
            }   
            --K;    
        }
        return map[dst] == INT_MAX ? -1 : map[dst];
    }
};
